### PR TITLE
Add e2e tests and a test suite

### PR DESCRIPTION
**Pull request by tomgeorge**
Added some unit tests and e2e tests.

Biggest refactor was to change `config` from using an `init` function to return a `Config` struct.  This was done for testability.

Also refactored `puller.go` to look for a `$KUBECONFIG`, `~/.kube/config`, then in cluster config, again for testing.

### CODE REVIEW

File: cfg/config_test.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/google/go-cmp/cmp"
+	"os"
+	"testing"
+)
+
+func TestEnvVars(t *testing.T) {
+
+	defer unsetEnv()
+
+	os.Setenv("IMAGES", "che-server=eclipse/che-server:nightly")
```
Comment (ibuziuk) [inline cfg/config_test.go]: maybe let's change this for che-theia (no one is going to cache che-server images)

File: cfg/config_test.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+			env: map[string]string{
+				"DAEMONSET_NAME": "custom-daemonset-name",
+				"NAMESPACE":      "my-namespace",
+				"NODE_SELECTOR":  "{\"type\": \"compute\"}",
+			},
+			want: Config{
+				DaemonsetName: "custom-daemonset-name",
+				Namespace:     "my-namespace",
+				Images: map[string]string{
+					"che-server": "eclipse/che-server:nightly",
```
Comment (ibuziuk) [inline cfg/config_test.go]: same, would propose to change it to che-theia

File: cico_build_nightly.sh [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+
+set -u
+set -e
+
+LOCAL_IMAGE_NAME='kubernetes-image-puller'
+REGISTRY='quay.io'
+ORGANIZATION='openshiftio'
```
Comment (ibuziuk) [inline cico_build_nightly.sh]: we should not push upstream images here

File: cico_build_nightly.sh [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+RHEL_IMAGE_NAME='rhel-kubernetes-image-puller'
```
Comment (ibuziuk) [inline cico_build_nightly.sh]: let's not build rhel image for upstream

Comment (tomgeorge) [reply cico_build_nightly.sh]: This will be fixed in #8

**Comment by tomgeorge** [comment]
I have made the suggested changes cc @ibuziuk

File: go.mod [hunk 5]
```diff
@@ -4,17 +4,16 @@ go 1.13
 
 require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/evanphx/json-patch v4.5.0+incompatible // indirect
```
Comment (amisevsk) [inline go.mod]: Maybe run `go mod tidy` to clean up dependencies.

File: Makefile [hunk 6]
```diff
@@ -6,9 +6,12 @@ all: build docker
 
 .PHONY: build docker
 
-build:
+build: test
 	GOOS=linux go build -v -o ./bin/${BINARY_NAME} ./cmd/main.go
 
+test:
+	go test -v ./cfg... ./pkg... ./utils...
+
```
Comment (amisevsk) [inline Makefile]: Consider also adding a `ci` target to run `golangci-lint` as we have for the plugin broker: [Makefile](https://github.com/eclipse/che-plugin-broker/blob/ac7f5c3fb78593444bd634be72b4ed42451e3f6b/Makefile#L10) and [Dockerfile](https://github.com/eclipse/che-plugin-broker/blob/master/build/CI/Dockerfile)

**Comment by tomgeorge** [comment]
I agree, I think in the future I will refactor to pass the config to some top level struct, which can be instantiated in tests and in the real app, and eliminate the need to set/unset the environment.

File: Makefile (hunk 6 above)
Comment (tomgeorge) [reply Makefile]: awesome, I will add that

File: go.mod (hunk 5 above)
Comment (tomgeorge) [reply go.mod]: Good catch, will do

**Pull request merged**